### ISSUE TITLE
yolov5-four-PANet

### ISSUE THREAD

**Issue by SpongeBab** [SpongeBab/yolov5#3]
## 🚀 Feature

**yolov5-PANet-fourhead**
```
# Parameters
nc: 80  # number of classes
depth_multiple: 1.0  # model depth multiple
width_multiple: 1.0  # layer channel multiple
anchors: 3

#  - [ 10,13, 16,30, 33,23 ]  # P3/8
#  - [ 30,61, 62,45, 59,119 ]  # P4/16
#  - [ 116,90, 156,198, 373,326 ]  # P5/32

#  - [ 5,7, 8,12, 15,13 ]  # P2/4  # custom
#  - [ 10,13, 16,30, 33,23 ]  # P3/8
#  - [ 30,61, 62,45, 59,119 ]  # P4/16
#  - [ 116,90, 156,198, 373,326 ]  # P5/32

# YOLOv5 backbone
backbone:
  # [from, number, module, args]
  [ [ -1, 1, Focus, [ 64, 3 ] ],  # 0-P1/2
    [ -1, 1, Conv, [ 128, 3, 2 ] ],  # 1-P2/4
    [ -1, 3, BottleneckCSP, [ 128 ] ],
    [ -1, 1, Conv, [ 256, 3, 2 ] ],  # 3-P3/8
    [ -1, 9, BottleneckCSP, [ 256 ] ],
    [ -1, 1, Conv, [ 512, 3, 2 ] ],  # 5-P4/16
    [ -1, 9, BottleneckCSP, [ 512 ] ],
    [ -1, 1, Conv, [ 1024, 3, 2 ] ],  # 7-P5/32
    [ -1, 1, SPP, [ 1024, [ 5, 9, 13 ] ] ],
    [ -1, 3, BottleneckCSP, [ 1024, False ] ],  # 9
  ]

# YOLOv5 PANet head
head:
  [ [ -1, 1, Conv, [ 512, 1, 1 ] ],
    [ -1, 1, nn.Upsample, [ None, 2, 'nearest' ] ],
    [ [ -1, 6 ], 1, Concat, [ 1 ] ],  # cat backbone P4
    [ -1, 3, BottleneckCSP, [ 512, False ] ],  # 13

    [ -1, 1, Conv, [ 256, 1, 1 ] ],
    [ -1, 1, nn.Upsample, [ None, 2, 'nearest' ] ],
    [ [ -1, 4 ], 1, Concat, [ 1 ] ],  # cat backbone P3
    [ -1, 3, BottleneckCSP, [ 256, False ] ],  # 17 (P3/8-small)

    [ -1, 1, Conv, [ 128, 1, 1 ] ],
    [ -1, 1, nn.Upsample, [ None, 2, 'nearest' ] ],
    [ [ -1, 2 ], 1, Concat, [ 1 ] ],  # cat backbone P2
    [ -1, 3, BottleneckCSP, [ 128, False ] ],  # 17 (P2/4-small)

    [ -1, 1, Conv, [ 128, 3, 2 ] ],
    [ [ -1, 18 ], 1, Concat, [ 1 ] ],  # cat head P3
    [ -1, 3, BottleneckCSP, [ 256, False ] ],  # 17 (P3/8-small)

    [ -1, 1, Conv, [ 256, 3, 2 ] ],
    [ [ -1, 14 ], 1, Concat, [ 1 ] ],  # cat head P4
    [ -1, 3, BottleneckCSP, [ 512, False ] ],  # 20 (P4/16-medium)

    [ -1, 1, Conv, [ 512, 3, 2 ] ],
    [ [ -1, 10 ], 1, Concat, [ 1 ] ],  # cat head P5
    [ -1, 3, BottleneckCSP, [ 1024, False ] ],  # 23 (P5/32-large)

    [ [ 14, 17, 20, 23 ], 1, Detect, [ nc, anchors ] ],  # Detect(P3, P4, P5)
  ]
```

@glenn-jocher Can you help me with this?yolov5-panet four head(p5).I modified the network as the format of P6. But AP is not normal. What went wrong?
```
     Epoch   gpu_mem       box       obj       cls     total    labels  img_size
   297/299     3.22G   0.01925    0.0256         0   0.04485       133       640: 100%|██████████| 117/117 [00:35<00:00,  3.30it/s]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 9/9 [00:01<00:00,  5.17it/s]
                 all         50       1532     0.0064     0.0627   0.000436    4.9e-05

     Epoch   gpu_mem       box       obj       cls     total    labels  img_size
   298/299     3.22G   0.01946   0.02482         0   0.04428        87       640: 100%|██████████| 117/117 [00:35<00:00,  3.28it/s]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 9/9 [00:01<00:00,  5.09it/s]
                 all         50       1532     0.0066     0.0646   0.000461   5.19e-05

     Epoch   gpu_mem       box       obj       cls     total    labels  img_size
   299/299     3.22G   0.01916   0.02602         0   0.04517        75       640: 100%|██████████| 117/117 [00:35<00:00,  3.31it/s]
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 9/9 [00:02<00:00,  3.40it/s]
                 all         50       1532    0.00629     0.0614    0.00042   4.61e-05
```

**Comment by glenn-jocher** [comment]
@SpongeBab I would start from an official P6 yaml and make small changes at a time. If you train after each change then it will be easy to isolate cause and effect.
https://github.com/ultralytics/yolov5/blob/master/models/hub/yolov5l6.yaml

**Comment by SpongeBab** [comment]
@glenn-jocher
Oh, I made it. But I saw P2 yaml first, and I fixed my yaml according to it, and started training successfully.
I forgot an important point, as the network gets deeper and deeper with more PANET layers, the detection layer gets deeper, so I should increase the index of the detection layer, not decrease it. Maybe I did not say it clearly, but I think you must understand very well. It's just my negligence. Thank you so much!

So I shoud change:
`    [ [ 14, 17, 20, 23 ], 1, Detect, [ nc, anchors ] ],  # Detect(P2, P3, P4, P5)`
to
`    [ [ 21, 24, 27, 30 ], 1, Detect, [ nc, anchors ] ],  # Detect(P2, P3, P4, P5)`

Just record the result:
```
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 25/25 [00:05<00:00,  4.83it/s]
val: Scanning '..\Dataset\desk_3_400\valid.cache' images and labels... 50 found, 0 missing, 0 empty, 0 corrupted: 100%|██████████| 50/50 [00:13<?, ?it/s]
                 all         50       1532      0.989      0.938      0.974      0.748
Speed: 0.3ms pre-process, 36.9ms inference, 3.0ms NMS per image at shape (2, 3, 640, 640)
```
**yolov5-panet**:          mAP 0.728, latency 31.4ms
**yolov5-fourPanet**:   mAP 0.748(+2%), latency 40.2ms(+8.8ms)
